internal/allocator/allocator.go

```diff
@@ -245,89 +247,240 @@ func (a *Allocator) Unassign(svc string) {
 	}
+	return poolIps
+}
 
-	ips := []net.IP{}
-	ipfamilySel := make(map[ipfamily.Family]bool)
+// SelectIPsForFamilyPolicy returns a slice of IPs from the given poolIPs, which are suitable for the
+// given service ipfamily and IPFamilyPolicy.
+func (a *Allocator) SelectIPsForFamilyAndPolicy(
+	poolIps PoolIps,
+	serviceIPFamily ipfamily.Family,
+	serviceIPFamilyPolicy v1.IPFamilyPolicy,
+) ([]net.IP, error) {
+	ipv4, hasIPv4 := poolIps[ipfamily.IPv4]
+	ipv6, hasIPv6 := poolIps[ipfamily.IPv6]
 
-	switch serviceIPFamily {
-	case ipfamily.DualStack:
-		ipfamilySel[ipfamily.IPv4], ipfamilySel[ipfamily.IPv6] = true, true
-	default:
-		ipfamilySel[serviceIPFamily] = true
+	switch serviceIPFamilyPolicy {
+	case v1.IPFamilyPolicySingleStack:
+		if ip, ok := poolIps[serviceIPFamily]; ok {
+			return []net.IP{ip}, nil
+		}
+	case v1.IPFamilyPolicyRequireDualStack:
+		if hasIPv4 && hasIPv6 {
+			return []net.IP{ipv4, ipv6}, nil
+		}
+	case v1.IPFamilyPolicyPreferDualStack:
+		if hasIPv4 && hasIPv6 {
+			return []net.IP{ipv4, ipv6}, nil
+		}
+		if hasIPv4 {
+			return []net.IP{ipv4}, nil
+		}
+		if hasIPv6 {
+			return []net.IP{ipv6}, nil
+		}
 	}
+	return nil, fmt.Errorf("no available IPs in pool for %s IPFamily", serviceIPFamily)
+}
 
-	for _, cidr := range pool.CIDR {
-		cidrIPFamily := ipfamily.ForCIDR(cidr)
-		if _, ok := ipfamilySel[cidrIPFamily]; !ok {
-			// Not the right ip-family
+// SelectPoolIPsForService returns the PoolIps correspoding with the most suitable pool for a serviceIPFamily.
+func (a *Allocator) SelectPoolIPsForService(
+	pools []*config.Pool,
+	svcKey string,
+	svc *v1.Service,
+	serviceIPFamily ipfamily.Family,
+	ports []Port,
+	sharingKey, backendKey string,
+) (PoolIps, error) {
+	var primaryIPPoolCandidates, secondaryIPPoolCandidates []PoolIps
+	primaryIPFamily := ipfamily.IPv6
+	secondaryIPFamily := ipfamily.IPv4
+	serviceIPFamilyPolicy := v1.IPFamilyPolicySingleStack
+	if svc.Spec.IPFamilyPolicy != nil {
+		serviceIPFamilyPolicy = *(svc.Spec.IPFamilyPolicy)
+	}
+	// If there's no specification of preferred ip family, ipv4 has higher priority.
+	if len(svc.Spec.IPFamilies) == 0 || svc.Spec.IPFamilies[0] == v1.IPv4Protocol {
+		primaryIPFamily = ipfamily.IPv4
+		secondaryIPFamily = ipfamily.IPv6
+	}
+	for _, pool := range pools {
+		poolIps := a.getFreeIPsFromPool(pool, svcKey, ports, sharingKey, backendKey)
+		if _, ok := poolIps[serviceIPFamily]; ok {
+			return poolIps, nil
+		}
+
+		_, hasPrimaryIP := poolIps[primaryIPFamily]
+		_, hasSecondaryIP := poolIps[secondaryIPFamily]
+
+		if hasPrimaryIP && hasSecondaryIP {
+			return poolIps, nil
+		}
+
+		if serviceIPFamilyPolicy != v1.IPFamilyPolicyPreferDualStack {
 			continue
 		}
-		ip := a.getIPFromCIDR(cidr, pool.AvoidBuggyIPs, svcKey, ports, sharingKey, backendKey)
-		if ip != nil {
-			ips = append(ips, ip)
-			delete(ipfamilySel, cidrIPFamily)
+		if hasPrimaryIP {
+			primaryIPPoolCandidates = append(primaryIPPoolCandidates, poolIps)
+		} else if hasSecondaryIP {
```

Comment:
The idea with the slice is to loop through all pools before deciding. It could be that one of the later pools has dual stack, and we shouldn't return the first one that has any of these condition true.